mlflow/data/dataset_registry.py

```diff
@@ -0,0 +1,155 @@
+                f" Constructor name must start with 'load_' or 'from_'.",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        signature = inspect.signature(constructor_fn)
+        parameters = signature.parameters
+        for expected_kwarg in ["name", "digest"]:
+            if expected_kwarg not in parameters or parameters[expected_kwarg].kind not in [
+                inspect.Parameter.KEYWORD_ONLY,
+                inspect.Parameter.POSITIONAL_OR_KEYWORD,
+            ]:
+                raise MlflowException(
+                    f"Invalid dataset constructor function: {constructor_fn.__name__}. Function"
+                    f" must define an optional parameter named '{expected_kwarg}'.",
+                    INVALID_PARAMETER_VALUE,
+                )
+
+        if not issubclass(signature.return_annotation, Dataset):
+            raise MlflowException(
+                f"Invalid dataset constructor function: {constructor_fn.__name__}. Function must"
+                f" have a return type annotation that is a subclass of"
+                f" :py:class:`mlflow.data.dataset.Dataset`.",
+                INVALID_PARAMETER_VALUE,
+            )
+
+
+def register_constructor(constructor_fn: callable, constructor_name: str = None) -> str:
+    """
+    Registers a dataset constructor.
+
+    :param constructor_fn: A function that accepts at least the following
+                           inputs and returns an instance of a subclass of
+                           :py:class:`mlflow.data.dataset.Dataset`:
+
+                           - name: Optional. A string dataset name
+                           - digest: Optional. A string dataset digest.
+
+    :param constructor_name: The name of the constructor, e.g.
+                             "from_spark". The name must begin with the
+                             string "from_" or "load_". If unspecified, the `__name__`
+                             attribute of the `constructor_fn` is used instead and must
+                             begin with the string "from_" or "load_".
+    :return: The name of the registered constructor, e.g. "from_pandas" or "load_delta".
+    """
+    registered_constructor_name = _dataset_registry.register_constructor(
+        constructor_fn=constructor_fn, constructor_name=constructor_name
+    )
+    setattr(mlflow.data, registered_constructor_name, constructor_fn)
+    mlflow.data.__all__.append(registered_constructor_name)
+    return registered_constructor_name
+
+
+def get_registered_constructors() -> Dict[str, callable]:
+    """
+    Obtains the registered dataset constructors.
+
+    :return: A dictionary mapping constructor names to constructor functions.
+    """
+    return _dataset_registry.constructors
+
+
+_dataset_registry = DatasetRegistry()
+_dataset_registry.register_entrypoints()
+try:
+    from mlflow.data.pandas_dataset import from_pandas
+
+    _dataset_registry.register_constructor(from_pandas)
+except ImportError:
+    pass
```

Comment:
We can use `contextlib.suppress` here: https://docs.python.org/3/library/contextlib.html#contextlib.suppress